### ISSUE TITLE
mod_wsgi windows 7 not cooperating

### ISSUE THREAD

**Issue by mickeyjf** [GrahamDumpleton/mod_wsgi#211]
I am trying to run a Python program from Apache on a Windows 7 machine. I don't have much experience with either Apache or Python. After a number of false starts and dead ends ('don't use cgi', no mod_wsgi.so for recent Windows, mod_wsgi-express.exe does not have all features for Windows version, etc.) I have ended up here. 

Python is in the path, and can be run from the command line.
Apache is 2.4.27 VC15, is listening on 8081, and is able to serve a "hello world" index.html located in C:/apache24/htdocs.

After running mod_wsgi-express.exe and copying the output to the very end of httpd.conf, httpd.conf contains:

-------------------------
LoadFile "c:/program files/python36/python36.dll"
LoadModule wsgi_module "c:/program files/python36/lib/site-packages/mod_wsgi/server/mod_wsgi.cp36-win_amd64.pyd"
WSGIPythonHome "c:/program files/python36"

WSGIScriptAlias /wsgi "C:/wsgi_app"
<Directory "C:/wsgi_app">
AllowOverride None
Options None
Order deny,allow
Allow from all
</Directory>

-------------------------

All "AddHandler" lines in httpd.conf have been commented out.
All other "ScriptAlias" (i.e. cgi) ines in httpd.conf have been commented out.

I have a test app, "wsgi_app.py" located in C:/wsgi_app and cloned to "wsgi_app.wsgi", containing the following :

-------------------------
def application(environ, start_response):
    status = '200 OK'
    output = 'This is my Website! TA DA!!'

    response_headers = [('Content-type', 'text/plain'),
                        ('Content-Length', str(len(output)))]
    start_response(status, response_headers)

    return [output]
-------------------------    
    
When I try any of 

localhost:8081/wsgi
localhost:8081/wsgi/wsgi_app.py
localhost:8081/wsgi/wsgi_app.wsgi

Or, from another PC on the LAN, 'name_of_this_pc':8081/wsgi

I get :

Forbidden
You don't have permission to access [path] on this server.

where [path] is /wsgi , /wsgi/wsgi_app.py, or /wsgi/wsgi_app.wsgi

Also, although I have neither started httpd as a service or started it from a command line, it serves my index file after I restart windows. If I make a change to the "Hello World" file and refresh the browser, I do see the change, so it is not just reading a cache. Using taskmanager, I am unable to find httpd (or "apache") either as an application, a process or a service. This means that I can't reliably know that I have restarted it unless I restart Windows itself.

I do see "api_test" active in services - but I'm not sure what relation that might have to this.

What (probably) simple thing am I missing here?

Thanks in advance for any help you are able to give!

**Comment by mickeyjf** [comment]
I see that the pid in httpd.pid in Logs is the same as the one shown for test_api in services, and when I look in the services listing it's description is "Apache/2.4.27 (Win64) mod_wsgi/4.5.17 Python/3.6". Apparently installing mod_wsgi automatically installs and runs httpd as a service? I was not expecting that. Is this mentioned in a document anywhere?

**Comment by mickeyjf** [comment]
By setting the Logon Property of the service to myself I no longer get permission errors, but I still am not able to run a wsgi_app.py. If I put that file in htdocs the contents are displayed, but when in the WSGIScriptAlias directory the URL is "not found".

**Comment by GrahamDumpleton** [comment]
Which Apache distribution for Windows are you using? I only recommend that from:

* https://www.apachelounge.com/

Next, the configuration of is better as:

```
WSGIScriptAlias / "C:/wsgi_app/wsgi_app.py"
<Directory "C:/wsgi_app">
Require all granted
</Directory>
```

Use ``WSGIScriptAlias`` to refer to the actual WSGI script file. That is better than doing it as a directory. Then access it as ``http://localhost:8081`` from same machine.

For Apache 2.4, you should use ``Require all granted`` not ``Order/Allow``.

Since you are using Python 3.6, the WSGI hello world program you should use is:

```
def application(environ, start_response):
     status = '200 OK'
     output = b'This is my Website! TA DA!!'

    response_headers = [('Content-type', 'text/plain'),
                                        ('Content-Length', str(len(output)))]
    start_response(status, response_headers)

    return [output]
```

The ``output`` needs to be bytes, not Unicode string.

Installing mod_wsgi doesn't set up Apache as a service. How you do that depends on what Apache distribution you are using, as does how you restart it. When using ApacheLounge distribution, from memory it gives you an icon down in dock tool bar for restarting it.

**Comment by mickeyjf** [comment]
Graham, thanks for your response. I have made those changes, and no longer get a "URL not found" but instead a 500 Internal Server Error. I am using the Apache Windows binary from Apache Lounge, but it did not install with a shortcut icon. I see that there is a test_api.py buried deep in C:\Program Files\Python36\Lib\test\test_importlib, so I suppose this must have been set up to be a service as part of the Python install (??). I have changed it from automatic to manual so that I can start and stop httpd either from a command line or as a service, but in any case this makes no difference to its behavior. 

Ah! The apache error log showed that the python app had a IndentationError. (From line 5 on) I fixed that and am now getting the expected response. As I said, I don't have a lot of experience with Python so blank space is not something I think about first. At this point I should be able to proceed and only need to worry about my own code being correct. Thanks again!
